// gherking/ElementProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementProcessor = void 0;
const tslib_1 = require("tslib");
const debug_1 = require("./debug");
const gherkin_ast_1 = require("gherkin-ast");
const BackgroundProcessor_1 = require("./BackgroundProcessor");
const Processor_1 = require("./Processor");
const ScenarioProcessor_1 = require("./ScenarioProcessor");
const ScenarioOutlineProcessor_1 = require("./ScenarioOutlineProcessor");
const debug = (0, debug_1.getDebugger)("ElementProcessor");
class ElementProcessor extends Processor_1.ListProcessor {
    constructor(preCompiler) {
        super(preCompiler);
        this.backgroundProcessor = new BackgroundProcessor_1.BackgroundProcessor(preCompiler);
        this.scenarioProcessor = new ScenarioProcessor_1.ScenarioProcessor(preCompiler);
        this.scenarioOutlineProcessor = new ScenarioOutlineProcessor_1.ScenarioOutlineProcessor(preCompiler);
    }
    preFilter(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("preFilter(e: %s, p: %s, i: %d)", e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            if (e instanceof gherkin_ast_1.Background) {
                debug("preFilter -> Background");
                return this.backgroundProcessor.preFilter(e, p);
            }
            if (e instanceof gherkin_ast_1.Scenario) {
                debug("preFilter -> Scenario");
                return this.scenarioProcessor.preFilter(e, p, i);
            }
            /* istanbul ignore else */
            if (e instanceof gherkin_ast_1.ScenarioOutline) {
                debug("preFilter -> ScenarioOutline");
                return this.scenarioOutlineProcessor.preFilter(e, p, i);
            }
        });
    }
    postFilter(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("postFilter(e: %s, p: %s, i: %d)", e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            if (e instanceof gherkin_ast_1.Background) {
                debug("postFilter -> Background");
                return this.backgroundProcessor.postFilter(e, p);
            }
            if (e instanceof gherkin_ast_1.Scenario) {
                debug("postFilter -> Scenario");
                return this.scenarioProcessor.postFilter(e, p, i);
            }
            /* istanbul ignore else */
            if (e instanceof gherkin_ast_1.ScenarioOutline) {
                debug("postFilter -> ScenarioOutline");
                return this.scenarioOutlineProcessor.postFilter(e, p, i);
            }
        });
    }
    process(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("process(e: %s, p: %s, i: %d)", e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            if (e instanceof gherkin_ast_1.Background) {
                debug("process -> Background");
                return this.backgroundProcessor.process(e, p);
            }
            if (e instanceof gherkin_ast_1.Scenario) {
                debug("process -> Scenario");
                return this.scenarioProcessor.process(e, p, i);
            }
            /* istanbul ignore else */
            if (e instanceof gherkin_ast_1.ScenarioOutline) {
                debug("process -> ScenarioOutline");
                return this.scenarioOutlineProcessor.process(e, p, i);
            }
        });
    }
}
exports.ElementProcessor = ElementProcessor;
