// gherking/ScenarioOutlineProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScenarioOutlineProcessor = void 0;
const tslib_1 = require("tslib");
const gherkin_ast_1 = require("gherkin-ast");
const ExamplesProcessor_1 = require("./ExamplesProcessor");
const Processor_1 = require("./Processor");
const StepProcessor_1 = require("./StepProcessor");
const TagProcessor_1 = require("./TagProcessor");
const debug_1 = require("./debug");
const debug = (0, debug_1.getDebugger)("ScenarioOutlineProcessor");
class ScenarioOutlineProcessor extends Processor_1.PartialListProcessor {
    constructor(preCompiler) {
        debug("constructor(%o)", preCompiler);
        super(preCompiler);
        this.stepProcessor = new StepProcessor_1.StepProcessor(preCompiler);
        this.tagProcessor = new TagProcessor_1.TagProcessor(preCompiler);
        this.examplesProcessor = new ExamplesProcessor_1.ExamplesProcessor(preCompiler);
    }
    preFilter(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("preFilter(hasPreScenarioOutline: %s, e: %s, p: %s, i: %d)", !!this.preCompiler.preScenarioOutline, e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            return !this.preCompiler.preScenarioOutline || (yield this.preCompiler.preScenarioOutline(e, p, i));
        });
    }
    postFilter(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("postFilter(hasPostScenarioOutline: %s, e: %s, p: %s, i: %d)", !!this.preCompiler.postScenarioOutline, e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            return !this.preCompiler.postScenarioOutline || (yield this.preCompiler.postScenarioOutline(e, p, i));
        });
    }
    process(e, p, i) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /* istanbul ignore next */
            debug("process(hasOnScenarioOutline: %s, e: %s, p: %s, i: %d)", !!this.preCompiler.onScenarioOutline, e === null || e === void 0 ? void 0 : e.constructor.name, p === null || p === void 0 ? void 0 : p.constructor.name, i);
            let scenarioOutlines = e;
            if (this.preCompiler.onScenarioOutline) {
                const result = yield this.preCompiler.onScenarioOutline(e, p, i);
                if (typeof result !== "undefined") {
                    scenarioOutlines = result;
                }
            }
            if (scenarioOutlines) {
                if (Array.isArray(scenarioOutlines)) {
                    debug("...Array: %d", scenarioOutlines.length);
                    scenarioOutlines = yield ScenarioOutlineProcessor.map(scenarioOutlines, this.postProcess.bind(this));
                }
                else {
                    debug("...replace");
                    scenarioOutlines = yield this.postProcess(scenarioOutlines);
                }
            }
            return scenarioOutlines;
        });
    }
    postProcess(e) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            debug("postProcess(tags: %s, steps: %s, examples: %s)", 
            // @ts-ignore
            Array.isArray(e.tags), Array.isArray(e.steps), Array.isArray(e === null || e === void 0 ? void 0 : e.examples));
            e.tags = yield this.tagProcessor.execute(e.tags, e);
            e.steps = yield this.stepProcessor.execute(e.steps, e);
            if (e instanceof gherkin_ast_1.ScenarioOutline && e.examples) {
                e.examples = yield this.examplesProcessor.execute(e.examples, e);
            }
            return e;
        });
    }
}
exports.ScenarioOutlineProcessor = ScenarioOutlineProcessor;
